overlord/devicestate/devicestate_remodel_test.go

```diff
@@ -1975,3 +1978,543 @@ func (s *deviceMgrRemodelSuite) TestRemodelUC20LabelConflictsError(c *C) {
+		// 4 - promote recovery system
+		// 5 - reseal with new model as current and try; before reboot
+		//     set-model changed the model in the state, the new model
+		//     replaced the old one, and thus the remodel context
+		//     carries the new model in ground context
+		// 6 - reseal with new model as current
+		c.Check(modeenv.GoodRecoverySystems, DeepEquals, []string{"0000", expectedLabel})
+		c.Check(modeenv.CurrentRecoverySystems, DeepEquals, []string{"0000", expectedLabel})
+		switch resealKeyCalls {
+		case 2:
+			c.Check(modeenv.Model, Equals, model.Model())
+			c.Check(modeenv.TryModel, Equals, new.Model())
+			c.Check(filepath.Join(boot.InitramfsUbuntuBootDir, "device/model"),
+				testutil.FileContains, fmt.Sprintf("model: %s\n", model.Model()))
+			// old model's revision is 0
+			c.Check(filepath.Join(boot.InitramfsUbuntuBootDir, "device/model"),
+				Not(testutil.FileContains), "revision:")
+		case 3:
+			c.Check(modeenv.Model, Equals, new.Model())
+			c.Check(modeenv.TryModel, Equals, "")
+			c.Check(filepath.Join(boot.InitramfsUbuntuBootDir, "device/model"),
+				testutil.FileContains, fmt.Sprintf("model: %s\n", new.Model()))
+			c.Check(filepath.Join(boot.InitramfsUbuntuBootDir, "device/model"),
+				testutil.FileContains, fmt.Sprintf("revision: %v\n", new.Revision()))
+		case 5:
+			c.Check(modeenv.Model, Equals, model.Model())
+			c.Check(modeenv.TryModel, Equals, new.Model())
+			// we are in an after reboot scenario, the file contains
+			// the new model
+			c.Check(filepath.Join(boot.InitramfsUbuntuBootDir, "device/model"),
+				testutil.FileContains, fmt.Sprintf("model: %s\n", new.Model()))
+			c.Check(filepath.Join(boot.InitramfsUbuntuBootDir, "device/model"),
+				testutil.FileContains, fmt.Sprintf("revision: %v\n", new.Revision()))
+		case 6:
+			c.Check(modeenv.Model, Equals, new.Model())
+			c.Check(modeenv.TryModel, Equals, "")
+			c.Check(filepath.Join(boot.InitramfsUbuntuBootDir, "device/model"),
+				testutil.FileContains, fmt.Sprintf("model: %s\n", new.Model()))
+			c.Check(filepath.Join(boot.InitramfsUbuntuBootDir, "device/model"),
+				testutil.FileContains, fmt.Sprintf("revision: %v\n", new.Revision()))
+		}
+		return nil
+	})
+	defer restore()
+
+	chg, err := devicestate.Remodel(s.state, new)
+	c.Assert(err, IsNil)
+
+	// since we cannot panic in random place in code that runs under
+	// taskrunner, we reset the task status and retry the change again, but
+	// we cannot do that once a change has become ready, thus inject a task
+	// that will request a reboot, thus stopping execution before the change
+	// is ready
+	fakeRebootCalls := 0
+	fakeRebootCallsReady := false
+	s.o.TaskRunner().AddHandler("fake-reboot", func(task *state.Task, _ *tomb.Tomb) error {
+		fakeRebootCalls++
+		if fakeRebootCalls == 1 {
+			st := task.State()
+			st.Lock()
+			defer st.Unlock()
+			st.RequestRestart(state.RestartSystemNow)
+		}
+		if fakeRebootCallsReady {
+			return nil
+		}
+		return &state.Retry{}
```

Comment:
also maybe a short comment before this one too:

// retry => keep the change not ready